### ISSUE TITLE
CLI history (Bash commands) does not persist across sessions which are not manually closed

### ISSUE THREAD

**Issue by Chealer** [docker/for-win#10393]
- [X] I have tried with the latest version of Docker Desktop
  - [X] I have tried disabling enabled experimental features
  - [ ] I have uploaded Diagnostics
  - Diagnostics ID:

### Actual behavior
The history of shell sessions which are not manually closed is lost. Technically, shells must be brutally killed, so .bash_history remains unchanged.

### Expected behavior
Shell history should be saved before the container is powered off. Technically, processes should be sent a signal for clean termination (SIGTERM).

### Information
I have been using Docker for several weeks and I never saw it remember history beyond the current session. So this issue is not specific to a single container, nor to version 3.1.0 (I believe version 2 behaved the same).

This is visibly perfectly reproducible.

  - Windows Version: 10
  - Docker Desktop Version: 3.1.0
  - Hyper-V backend
  - Are you running inside a virtualized Windows e.g. on a cloud server or a VM: No

The obvious workaround for this is to manually close shells. But even manually closing by **clicking the "X" widget or with Alt+F4 will not suffice**. For some reason, this seems also broken, so to keep history, it is necessary to use the shell's _exit_ builtin (a single invocation per Bash session suffices, even if Bash runs inside a parent shell).

For GNU Bash, there is a more realistic workaround which is to use [a brilliant hack to save history continuously](https://askubuntu.com/questions/67283/is-it-possible-to-make-writing-to-bash-history-immediate) (altering the PROMPT_COMMAND variable).

### Steps to reproduce the behavior
  1. Start a Linux container which has a shell which keeps history such as GNU Bash (for example, [Ubuntu](https://hub.docker.com/search?q=ubuntu&source=verified))
  2. Open a shell in the container using the container's CLI button
  3. Launch a shell with history such as GNU Bash
  4. Stop the container

### Steps to note the behavior

1. (Re)start the container
  2. Open a shell in the container using the container's CLI button
  3. Launch the same shell with history
  4. Display the history, for example using $ history

**Comment by stephen-turner** [comment]
In my experiments, `exit` from within the bash shell exits cleanly and preserves the history. `docker stop` from outside the container kills the container.

**Comment by stephen-turner** [comment]
Thanks for the report, @Chealer. From the GUI, we launch the shell with a `docker exec`. But `docker stop` only sends SIGTERM to the container's original process (the one specified in the `docker run`): each `docker exec` is spawned as a new shell. (In this respect, the problem with the history is similar to the problem of maintaining history across two shells).

If instead of pressing the CLI button in the GUI, you open a PowerShell terminal and use `docker attach` to connect to the container's original shell, it will have the behaviour you want. But we can't do that in the GUI because we don't know what program the container was launched with.

In summary, this would have to be fixed upstream of us, in the engine. moby/moby#9098 is closely related and there's an open PR at moby/moby#41548 too.

**Comment by Chealer** [comment]
@stephen-turner I am not sure what problem of maintaining history across two shells you are referring to, but thank you very much for the explanation
Glad to see it's being worked on

**Comment by Chealer** [comment]
Why is this ticket closed?

**Comment by stephen-turner** [comment]
It's closed because it's not in our codebase, it's upstream of us.

**Comment by Chealer** [comment]
Thanks @stephen-turner but closed tickets are not those about upstream issues. Tickets are closed when the issues they track are resolved.

**Comment by Chealer** [comment]
Can this ticket be reopened? I'm afraid this persists.

**Comment by stephen-turner** [comment]
As I have explained, we do not hold tickets open for upstream issues, only bugs in our codebase, otherwise we would have an unlimited number of tickets we have no control over. Once it is fixed upstream, it will automatically make its way to us in the next update.
